src/commands/index.js

```diff
@@ -0,0 +1,171 @@
+// @ts-check
+
+const { userBlacklist, commandPrefix } = require('../config/config')
+const { groupMemberships, groups } = require('../config/config2.js')
+const { getAllFiles } = require('../utils/file.js')
+
+class CommandManager {
+  /**
+   * @type {import('../controller')}
+   */
+  #controller
+
+  /**
+   * @type {Record<string, import('./types.d.ts').Command>}
+   */
+  #commands = {}
+
+  /**
+   * @param {import('../controller')} controller 
+   */
+  constructor(controller) {
+    if (!controller.connections.twitch) {
+      throw new TypeError('controller.connections.twitch not found')
+    }
+
+    this.#controller = controller
+
+    controller.connections.twitch.onMessage(this.#handleTwitchMessage)
+  }
+
+  async loadCommands() {
+    const files = (await getAllFiles(__dirname)).filter(path => path.endsWith('.js') && path !== 'index.js')
+
+    for (const path of files) {
+      const module = require(path)
+
+      const commands = Array.isArray(module) ? module : [module]
+      for (const command of commands) {
+        assertCommand(path, command)
+
+        this.#commands[command.name] = command
+      }
+    }
+  }
+
+  /**
+   * @param {string} channel 
+   * @param {string} user 
+   * @param {string} text 
+   * @param {object} msg TODO grab type
+   * 
+   * @returns {Promise<void>}
+   */
+  async #handleTwitchMessage(channel, user, text, msg) {
+    text = text.trim()
+
+    if (text.length < 2) {
+      // Not enough for their to be a command
+      return;
+    }
+
+    if (userBlacklist.includes(user.toLowerCase())) {
+      return;
+    }
+
+    const args = text.split(' ')
+    if (!args[0].startsWith(commandPrefix)) {
+      // Not a command (that we care about)
+      return;
+    }
+
+    const commandName = args[0].substring(1)
+    const command = this.#commands[commandName]
+    if (!command || !command.enabled) {
+      // Command doesn't exist or it's disabled
+      return;
+    }
+
+    if (!canUserPerformCommand(user, command)) {
+      return
+    }
+
+    const result = command.run({
+      controller: this.#controller,
+      channel,
+      user,
+      args,
+      msg
+    })
+
+    if (typeof result?.then === 'function') {
+      await result
+    }
```

Comment:
Instead of this if, I _think_ you could do `await Promise.resolve(command.run(...))`?